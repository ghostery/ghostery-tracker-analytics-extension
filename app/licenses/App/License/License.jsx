/**
 * @namespace LicenseClasses
 */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import './License.scss';

/**
 * @class Handles license entry on internal licenses.html page
 * which displays licenses for all third-party packages used by Ghostery
 * @memberOf  LicenseClasses
 */
class License extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expanded: false,
    };
  }

  /**
   * Toggle expansion of a license full text.
   */
  toggleLicenseText = () => {
    const { expanded } = this.state;
    this.setState({ expanded: !expanded });
  }

  /**
   * Render single license entry.
   * @return {ReactComponent}   ReactComponent instance
   */
  render() {
    const { license, isLastItem } = this.props;
    const {
      name,
      repository,
      licenses,
      publisher,
      url,
      email,
      licenseText,
    } = license;
    const { expanded } = this.state;

    const hasInfo = {
      name: typeof (name) !== 'undefined',
      repository: typeof (repository) !== 'undefined',
      licenses: typeof (licenses) !== 'undefined',
      publisher: typeof (publisher) !== 'undefined',
      url: typeof (url) !== 'undefined',
      email: typeof (email) !== 'undefined',
      licenseText: typeof (licenseText) !== 'undefined',
    };

    return (
      <Fragment>
        <div className="License">
          {hasInfo.name && <div>{`Module: ${name}`}</div>}
          {hasInfo.repository && (
            <div>
              <span>Repository: </span>
              <a className="License__link" href={repository} target="_blank" rel="noopener noreferrer">
                <span>
                  {repository}
                </span>
              </a>
            </div>
          )}
          {hasInfo.licenses && <div>{`License type(s): ${licenses}`}</div>}
          {hasInfo.publisher && <div>{`Publisher: ${publisher}`}</div>}
          {hasInfo.url && <div>{`Url: ${url}`}</div>}
          {hasInfo.email && <div>{`Email: ${email}`}</div>}
          {hasInfo.licenseText && (
            <div>
              <div className="License__textToggle" onClick={this.toggleLicenseText}>
                License Text
              </div>
              {
                expanded && (
                  <div className="License__text">
                    {licenseText}
                  </div>
                )
}
            </div>
          )}
        </div>
        {!isLastItem && <div className="License__hr" />}
      </Fragment>
    );
  }
}

License.propTypes = {
  license: PropTypes.shape({
    name: PropTypes.string.isRequired,
    repository: PropTypes.string,
    licenses: PropTypes.string.isRequired,
    publisher: PropTypes.string,
    email: PropTypes.string,
    licenseText: PropTypes.string,
    url: PropTypes.string,
  }).isRequired,
  isLastItem: PropTypes.bool.isRequired,
};

export default License;
