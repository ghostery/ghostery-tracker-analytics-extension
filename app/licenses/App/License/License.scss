@import '../../../utils/scss/open-sans.scss';
@import '../../../utils/scss/color-variables.scss';

.License {
  margin: 29px 0 29px 48px;
  font-family: OpenSans;
  font-weight: 400;
  font-size: 14px;
}

.License__link {
  text-decoration: none;
  color: $ghosty-blue;
}

.License__textToggle {
  font-weight: 700;
  cursor: pointer;
}

.License__text {
  margin: 6px 48px 0 0;
  border: 1px solid $white;
  padding: 5px;
  display: flex;
}

.License__hr {
  margin: 0 22px;
  border-bottom: 1px solid $medium-grey;
  display: flex;
}
