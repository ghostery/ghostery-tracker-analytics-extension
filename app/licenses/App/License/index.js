import License from './License.jsx';

export default License;
