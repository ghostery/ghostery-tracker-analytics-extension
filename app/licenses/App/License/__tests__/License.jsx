import React from 'react';
import { shallow } from 'enzyme';

import License from '../License';

describe('app/licenses/App/License', () => {
  describe('Snapshot tests with react-test-renderer', () => {
    test('License component should render', () => {
      const data = {
        license: {
          name: 'test',
          repository: 'test-repo',
          licenses: 'test-license',
          publisher: 'test-publisher',
          email: 'test@test.com',
          licenseText: 'test',
          url: 'tst.com',
        },
        isLastItem: false,
      };

      const component = shallow(<License {...data} />);
      expect(component).toMatchSnapshot();
    });
  });
});
