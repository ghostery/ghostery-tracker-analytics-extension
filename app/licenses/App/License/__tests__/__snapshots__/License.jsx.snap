// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/licenses/App/License Snapshot tests with react-test-renderer License component should render 1`] = `
<Fragment>
  <div
    className="License"
  >
    <div>
      Module: test
    </div>
    <div>
      <span>
        Repository: 
      </span>
      <a
        className="License__link"
        href="test-repo"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          test-repo
        </span>
      </a>
    </div>
    <div>
      License type(s): test-license
    </div>
    <div>
      Publisher: test-publisher
    </div>
    <div>
      Url: tst.com
    </div>
    <div>
      Email: test@test.com
    </div>
    <div>
      <div
        className="License__textToggle"
        onClick={[Function]}
      >
        License Text
      </div>
    </div>
  </div>
  <div
    className="License__hr"
  />
</Fragment>
`;
